pytest_datadir/plugin.py

```diff
@@ -25,24 +30,149 @@ def _win32_longpath(path):
+        exists = True
+        if not original_path.exists():
+            # raise the flag that it doesn't exist so we can generate
+            # a directory for it instead of copying
+            exists = False
+
+        # make sure the path is a directory if it exists
+        elif not original_path.is_dir():
+            raise ValueError("datadir path is not a directory")
+
+        # generate a base temporary directory and receive the path to it
+        temp_path = self.tmp_path_factory.mktemp('')
+
+        # in order to use the shutil.copytree util the target directory
+        # must not exist so we specify a dir in the generated tempdir for it
+        temp_data_path = temp_path / DATADIR_DIRNAME
+
+        # windows-ify the paths
+        original_path = Path(_win32_longpath(original_path))
+        temp_data_path = Path(_win32_longpath(str(temp_data_path)))
+
+        # copy or create empty directory depending on whether the
+        # original one exists
+        if exists:
+
+            # copy all the files in the original data dir to the temp
+            # dir
+            shutil.copytree(original_path, temp_data_path)
+
+        else:
+            # otherwise just give them a fallback tmpdir
+            temp_data_path.mkdir()
+
+        return temp_data_path
+
+# to clean up for the different scopes we need tmp_path_factory to be
+# scoped differently; just copy the code from module and change the
+# decorator
+
+@pytest.fixture(scope="session")
+def session_tmp_path_factory(request):
+    """Return a :class:`_pytest.tmpdir.TempPathFactory` instance for the test session.
+    """
+    return request.config._tmp_path_factory
+
+@pytest.fixture(scope="module")
+def module_tmp_path_factory(request):
+    """Return a :class:`_pytest.tmpdir.TempPathFactory` instance for the test session.
+    """
+    return request.config._tmp_path_factory
+
+
+@pytest.fixture(scope="class")
+def class_tmp_path_factory(request):
+    """Return a :class:`_pytest.tmpdir.TempPathFactory` instance for the test session.
+    """
+    return request.config._tmp_path_factory
+
+@pytest.fixture(scope="function")
+def function_tmp_path_factory(request):
+    """Return a :class:`_pytest.tmpdir.TempPathFactory` instance for the test session.
+    """
+    return request.config._tmp_path_factory
+
+# scoped factories
+@pytest.fixture(scope='module')
+def module_datadir_factory(request, module_tmp_path_factory):
+
+    return DatadirFactory(request, module_tmp_path_factory)
+
+@pytest.fixture(scope='class')
+def class_datadir_factory(request, class_tmp_path_factory):
+
+    return DatadirFactory(request, class_tmp_path_factory)
+
+@pytest.fixture(scope='function')
+def function_datadir_factory(request, function_tmp_path_factory):
+
+    return DatadirFactory(request, function_tmp_path_factory)
+
+
+# shared datadirs
+
+@pytest.fixture(scope='module')
+def module_shared_datadir(request, module_datadir_factory):
```

Comment:
I'm not sure having specific "shared_datadir" fixtures for each scope makes sense. Users could just use `datadir_factory` directly, similar to how pytest does it.